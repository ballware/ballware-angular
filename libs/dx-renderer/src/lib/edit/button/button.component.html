<dx-button *ngIf="livecycle.layoutItem && livecycle.layoutItem.options && livecycle.layoutItem.options.dataMember"  
  class="mt-2"
  [visible]="(visible.visible$ | async) ?? false"
  [text]="livecycle.layoutItem.options.hint ?? ''"
  [disabled]="(readonly.readonly$ | async) ?? false"
  [height]="livecycle.layoutItem.options.height"
  [width]="livecycle.layoutItem.options.width"
  (onClick)="livecycle.onEvent('click')">
</dx-button>
