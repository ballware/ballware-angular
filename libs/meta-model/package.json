{
  "name": "@ballware/meta-model",
  "version": "8.0.1",
  "type": "commonjs"
}
