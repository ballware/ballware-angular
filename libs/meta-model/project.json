{
  "name": "meta-model",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/meta-model/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": [
        "{options.outputPath}"
      ],
      "options": {
        "outputPath": "dist/libs/meta-model",
        "main": "libs/meta-model/src/index.ts",
        "tsConfig": "libs/meta-model/tsconfig.lib.json",
        "assets": [
          "libs/meta-model/*.md"
        ]
      }
    },
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs meta-model {args.ver} {args.tag}"
      },
      "dependsOn": [
        "build"
      ]
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": [
        "{options.outputFile}"
      ],
      "options": {
        "lintFilePatterns": [
          "libs/meta-model/**/*.ts"
        ]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": [
        "{workspaceRoot}/coverage/{projectRoot}"
      ],
      "options": {
        "jestConfig": "libs/meta-model/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
