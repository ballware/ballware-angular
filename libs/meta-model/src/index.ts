export * from './lib/cruditem';
export * from './lib/document';
export * from './lib/entity';
export * from './lib/page';
export * from './lib/queryparams';
export * from './lib/rights';
export * from './lib/scriptutil';
export * from './lib/statistic';
export * from './lib/template';
export * from './lib/tenant';

