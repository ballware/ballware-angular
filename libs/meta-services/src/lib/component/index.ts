import { StoreModule } from '@ngrx/store';
import { componentReducer } from './component.reducer';
import { componentFeatureKey } from './component.state';

export * from './component.actions';
export * from './component.state';

export const ComponentFeatureModule = StoreModule.forFeature(componentFeatureKey, componentReducer);