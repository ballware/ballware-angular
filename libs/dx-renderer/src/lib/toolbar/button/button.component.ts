import { AfterViewInit, Component, Input, OnDestroy, ViewChild } from '@angular/core';
import { PageService, ToolbarItemRef } from '@ballware/meta-services';
import { PageToolbarItem } from '@ballware/meta-model';
import { DxButtonComponent } from 'devextreme-angular';

@Component({
  selector: 'ballware-toolbar-button',
  templateUrl: './button.component.html',
  styleUrls: ['./button.component.scss']
})
export class ToolbarButtonComponent implements OnDestroy, AfterViewInit {

  @Input() public item?: PageToolbarItem;

  @ViewChild('button', { static: false }) button?: DxButtonComponent;

  constructor(private pageService: PageService) { }

  ngOnDestroy(): void {
    if (this.item?.name) {
      this.pageService.paramEditorDestroyed(this.item.name);
    }
  }

  ngAfterViewInit(): void {

    if (this.item?.name) {
      const toolbarItemRef = {
        getOption: (option) => this.button?.instance.option(option),
        setOption: (option, value) => this.button?.instance.option(option, value)
      } as ToolbarItemRef;

      this.pageService.paramEditorInitialized({ name: this.item.name, item: toolbarItemRef });
    }
  }

  onClicked(e: unknown) {
    if (this.item?.name) {
      this.pageService.paramEditorEvent({ name: this.item.name, event: 'click' });
    }
  }
}
