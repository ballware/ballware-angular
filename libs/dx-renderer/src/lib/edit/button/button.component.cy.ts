import { MountConfig, mount } from 'cypress/angular';
import { EditLayoutButtonComponent } from './button.component';

describe(EditLayoutButtonComponent.name, () => {
  const config: MountConfig<EditLayoutButtonComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(EditLayoutButtonComponent, config);
  })
})
