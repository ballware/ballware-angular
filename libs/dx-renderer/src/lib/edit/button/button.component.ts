import { Component, Inject, Input, OnInit } from '@angular/core';
import { EditLayoutItem } from '@ballware/meta-model';
import { EDIT_SERVICE, EditService } from '@ballware/meta-services';
import { ClickEvent } from 'devextreme/ui/button';
import { combineLatest, takeUntil } from 'rxjs';
import { WithDestroy } from '../../utils/withdestroy';
import { WithEditItemLifecycle } from '../../utils/withedititemlivecycle';
import { WithReadonly } from '../../utils/withreadonly';
import { WithVisible } from '../../utils/withvisible';
import { CommonModule } from '@angular/common';
import { DxButtonModule } from 'devextreme-angular';

@Component({
  selector: 'ballware-edit-button',
  templateUrl: './button.component.html',
  styleUrls: ['./button.component.scss'],
  imports: [CommonModule, DxButtonModule],
  standalone: true
})
export class EditLayoutButtonComponent extends WithVisible(WithReadonly(WithEditItemLifecycle(WithDestroy()))) implements OnInit {

  @Input() initialLayoutItem?: EditLayoutItem;

  public layoutItem: EditLayoutItem|undefined;

  private dataMember: string|undefined;
  private editorEvent: ((dataMember: string, event: string) => void)|undefined;

  constructor(
    @Inject(EDIT_SERVICE) private editService: EditService) {
    super();
  }

  ngOnInit(): void {
    if (this.initialLayoutItem) {
      this.initLifecycle(this.initialLayoutItem, this.editService, this);

      this.preparedLayoutItem$
        .pipe(takeUntil(this.destroy$))
        .subscribe((layoutItem) => {
          if (layoutItem) {
            this.initReadonly(layoutItem, this.editService);
            this.initVisible(layoutItem);

            this.layoutItem = layoutItem;            
            this.dataMember = layoutItem.options?.dataMember;
          }
        });

      combineLatest([this.editService.editorEvent$])
        .pipe(takeUntil(this.destroy$))
        .subscribe(([editorEvent]) => {
            if (editorEvent) {
              this.editorEvent = (dataMember, event) => editorEvent({ dataMember, event });
            }            
        });
    }
  }

  onClick(e: ClickEvent) {
    if (this.editorEvent && this.dataMember) {
      this.editorEvent(this.dataMember, 'click');
    }
  }
}
