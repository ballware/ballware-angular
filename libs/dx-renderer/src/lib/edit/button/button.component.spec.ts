import { ComponentFixture, TestBed } from '@angular/core/testing';

import { EditLayoutButtonComponent } from './button.component';
import { Provider } from '@angular/core';
import { EDIT_SERVICE } from '@ballware/meta-services';
import { EditLayoutItem } from '@ballware/meta-model';
import { mockedEditServiceContext } from '../../../test/editservice.spec';

import { ClickEvent } from 'devextreme/ui/button';

describe('EditLayoutButtonComponent', () => {
  let component: EditLayoutButtonComponent;
  let fixture: ComponentFixture<EditLayoutButtonComponent>;

  const mockedEditService = mockedEditServiceContext();
        
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ EditLayoutButtonComponent ],
      providers: [        
        {
          provide: EDIT_SERVICE,
          useFactory: () => mockedEditService.mock.object()
        } as Provider         
      ]
    })
    .compileComponents();
  });

  it('should create', () => {
    fixture = TestBed.createComponent(EditLayoutButtonComponent);

    const layoutItem = {
        options: {
            dataMember: 'mockedmember'
        }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    component.initialLayoutItem = layoutItem;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should apply options', () => {
    fixture = TestBed.createComponent(EditLayoutButtonComponent);

    const layoutItem = {
        options: {
            dataMember: 'mockedmember',
            readonly: false,            
            visible: false
        }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    component.initialLayoutItem = layoutItem;
    fixture.detectChanges();

    expect(component.getOption('readonly')).toBe(false);
    expect(component.getOption('visible')).toBe(false);
    
    component.setOption('readonly', true);
    expect(component.getOption('readonly')).toBe(true);

    component.setOption('visible', true);
    expect(component.getOption('visible')).toBe(true);

    expect(() => component.getOption('undefined')).toThrowError('Unsupported option <undefined>');
    expect(() => component.setOption('undefined', 'any value')).toThrowError('Unsupported option <undefined>');
  });

  it('should trigger click event', () => {    
    fixture = TestBed.createComponent(EditLayoutButtonComponent);

    const layoutItem = {
        options: {
            dataMember: 'mockedmember',
            readonly: false,            
            visible: true
        }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    component.initialLayoutItem = layoutItem;
    fixture.detectChanges();

    component.onClick({} as ClickEvent);

    expect(mockedEditService.editorEvent).toBeCalledTimes(1);
    expect(mockedEditService.editorEvent).toBeCalledWith({ dataMember: 'mockedmember', event: 'click'});
  });
});
