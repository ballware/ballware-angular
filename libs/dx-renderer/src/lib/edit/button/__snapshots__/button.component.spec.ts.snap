// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditLayoutButtonComponent should create 1`] = `
<ballware-edit-button
  _optionRegistry={[Function Array]}
  dataMember={[Function String]}
  destroy$={[Function Subject]}
  editService={[Function ]}
  editorEntered$={[Function Subject]}
  editorEvent={[Function Function]}
  editorEvent$={[Function Subject]}
  getOption={[Function Function]}
  globalReadonly="false"
  initialLayoutItem={[Function Object]}
  layoutItem={[Function Object]}
  onEntered={[Function Function]}
  onEvent={[Function Function]}
  preparedLayoutItem$={[Function BehaviorSubject]}
  readonly$={[Function BehaviorSubject]}
  registerOption={[Function Function]}
  setOption={[Function Function]}
  visible$={[Function BehaviorSubject]}
>
  <dx-button
    class="dx-widget dx-button dx-button-mode-contained dx-button-normal mt-2"
    role="button"
    tabindex="0"
  >
    <div
      class="dx-button-content"
    />
  </dx-button>
</ballware-edit-button>
`;
